x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilterTests.java

```diff
@@ -331,16 +334,31 @@ private static BulkItemRequest[] randomBulkItemRequest(
         for (var entry : fieldInferenceMap.values()) {
             String field = entry.getName();
             var model = modelMap.get(entry.getInferenceId());
-            String text = randomAlphaOfLengthBetween(10, 20);
-            docMap.put(field, text);
-            expectedDocMap.put(field, text);
+            Object inputObject = randomInputCasesForSemanticText();
+            String inputText = inputObject.toString();
+            docMap.put(field, inputObject);
+            expectedDocMap.put(field, inputText);
             if (model == null) {
                 // ignore results, the doc should fail with a resource not found exception
                 continue;
             }
-            var result = randomSemanticText(field, model, List.of(text), requestContentType);
-            model.putResult(text, toChunkedResult(result));
-            expectedDocMap.put(field, result);
+
+            SemanticTextField semanticTextField;
+            if (model.hasResult(inputText)) {
+                ChunkedInferenceServiceResults results = model.getResults(inputText);
+                semanticTextField = semanticTextFieldFromChunkedInferenceResults(
+                    field,
+                    model,
+                    List.of(inputText),
+                    results,
+                    requestContentType
+                );
+            } else {
+                semanticTextField = randomSemanticText(field, model, List.of(inputText), requestContentType);
+                model.putResult(inputText, toChunkedResult(semanticTextField));
+            }
```

Comment:
@carlosdelest We had to make this change because the inference result cache in `model` is not field-aware. Now that our input can be of many data types (including `Boolean`, with only two values), we are nearly guaranteed to hit value collisions across 100+ bulk requests. This caused test failures with the previous logic because different random embeddings would be generated every time we saw the value "true" (for example).

This updated logic checks if the inference result cache already has results for the value, and uses them if it does.

Reply:
I see - we could maybe have used `ESTestCase.randomValueOtherThanMany()`to a similar effect. That would ensure that the actual random value is not selected - AFAIU we should loop until we find a value that is not on the results?

Reply:
Maybe I'm misunderstanding your comment, but I don't think `ESTestCase.randomValueOtherThanMany()` would help here. The issue is that the previous logic _always_ generated a new embedding for the input, regardless of whether the model already had a cached value for that input. This caused test failures. Consider the following case:

- We generate a random embedding for the input `true`
- We write that embedding to the expected doc map and cache it in `model`
- In a later bulk request, the input `true` is randomly generated again
- We generate a _different_ random embedding for the input `true`
- We overwrite the cached embedding in `model` with the new embedding
- After all requests are generated, we assert that the embedding in the expected doc map matches that in the `model` cache. This fails because the embedding in the `model` cache was overwritten.

This new logic fixes the problem by first checking if `model` already has a cached embedding for the input. If it does, we use it. If it doesn't, we generate a new random embedding and add it to the `model` cache.

Reply:
My point is, wouldn't it be simpler not to generate the duplicate input values, and thus avoid managing the results as it happens?